zapcore/hook.go

```diff
@@ -66,3 +81,32 @@ func (h *hooked) Write(ent Entry, _ []Field) error {
 	}
 	return err
 }
+
+func (h *hookedWithFields) Check(ent Entry, ce *CheckedEntry) *CheckedEntry {
+	// Let the wrapped Core decide whether to log this message or not. This
+	// also gives the downstream a chance to register itself directly with the
+	// CheckedEntry.
+	if downstream := h.Core.Check(ent, ce); downstream != nil {
+		return downstream.AddCore(ent, h)
+	}
+	return ce
+}
+
+func (h *hookedWithFields) With(fields []Field) Core {
+	return &hookedWithFields{
+		Core:   h.Core.With(fields),
+		funcs:  h.funcs,
+		fields: append(h.fields[:len(h.fields):len(h.fields)], fields...),
```

Comment:
Couple of requests:
 * Can you move the creation of the new `fields` up and add comments
 * Can you add a unit test that verifies that the behaviour works correctly. Ideally, a test that fails with `append(h.fields, fields...)` but passes with the correct copy behaviour
* For performance reasons, I think using `append` could end up creating a slice larger than necessary.

See example: https://play.golang.org/p/MBNMAMzOsBs

The resulting slice may have `len < cap`, but this slice is never appended to (it's immutable after creation) -- we only ever make a full copy and add new fields.

It would be more efficient to do something like:
```go
copyFields := make([]Field, len(h.fields) + len(fields))
copy(copyFields, h.fields)
copy(copyFields[len(h.fields):], fields)
```

that way the capacity == length

Reply:
Don't know if I'm getting it right, so I'm gonna ask before commiting. Is the request to implement a separate method for creating new `fields` and then add a unit test for that?

Reply:
No need to create a separate method for appending the fields, just a separate test to verify that the original core's fields backing array is not shared by the new resulting fields. 

E.g., if this code was just `append(h.fields, fields...)`, then the backing fields array could be shared, and so a test should detect this misuse.